src/Raven.Server/Documents/Indexes/Index.cs

```diff
@@ -752,20 +768,40 @@ protected void ExecuteIndexing()
                             {
                                 try
                                 {
-                                    _cts.Token.ThrowIfCancellationRequested();
+                                    _indexingProcessCancellationTokenSource.Token.ThrowIfCancellationRequested();
+                                    _batchProcessCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(_indexingProcessCancellationTokenSource.Token);
 
-                                    bool didWork;
+                                    bool didWork = false;
                                     try
                                     {
                                         TimeSpentIndexing.Start();
-                                        didWork = DoIndexingWork(scope, _cts.Token);
+                                        didWork = DoIndexingWork(scope, _batchProcessCancellationTokenSource.Token);
+                                    }
+                                    catch (OperationCanceledException)
+                                    {
+                                        // We are here because the batch has been cancelled. This may happen
+                                        // because the database is shutting down, or the batch has been
+                                        // cancelled by some other internal process. If the database is
+                                        // shutting down, then we throw again to handle shutdown.
+                                        if (_indexingProcessCancellationTokenSource.IsCancellationRequested)
+                                        {
+                                            throw;
+                                        }
                                     }
                                     finally
                                     {
                                         TimeSpentIndexing.Stop();
+
+                                        // If we are here, then the previous block did not throw. There's two
+                                        // possibilities: either the batch was not cancelled, and we really
+                                        // did finish the work, or the batch was cancelled, but not because of
+                                        // a database shutdown
+                                        batchCompleted = !_batchProcessCancellationTokenSource.IsCancellationRequested;
+                                        _batchProcessCancellationTokenSource = null;
                                     }
 
-                                    _indexingBatchCompleted.SetAndResetAtomically();
+                                    if (batchCompleted)
```

Comment:
Because it may go through that flow even if the batch is not completed, and we don't want to let them know if we are just adjusting the memory usage because of a low memory notification, as it is just an internal process.

Should I change this back to the previous way? waiters would get notified even in low memory notification cases.